packages/snaps-controllers/src/websocket/WebSocketService.ts

```diff
@@ -0,0 +1,254 @@
+
+    this.#messenger.registerActionHandler(
+      `${serviceName}:sendMessage`,
+      this.sendMessage.bind(this),
+    );
+
+    this.#messenger.registerActionHandler(
+      `${serviceName}:getAll`,
+      this.getAll.bind(this),
+    );
+
+    this.#messenger.subscribe('SnapController:snapUpdated', (snap) => {
+      this.closeAll(snap.id);
+    });
+
+    this.#messenger.subscribe('SnapController:snapUninstalled', (snap) => {
+      this.closeAll(snap.id);
+    });
+
+    // Due to local Snaps not currently triggering uninstalled we also close connections for new Snaps.
+    this.#messenger.subscribe('SnapController:snapInstalled', (snap) => {
+      this.closeAll(snap.id);
+    });
+  }
+
+  #get(snapId: SnapId, id: string) {
+    const socket = this.#sockets.get(id);
+
+    assert(
+      socket && socket.snapId === snapId,
+      `Socket with ID "${id}" not found.`,
+    );
+
+    return socket;
+  }
+
+  #exists(snapId: SnapId, url: string) {
+    return this.getAll(snapId).some((socket) => socket.url === url);
+  }
+
+  #handleEvent(snapId: SnapId, event: WebSocketEvent) {
+    this.#messenger
+      .call('SnapController:handleRequest', {
+        origin: METAMASK_ORIGIN,
+        snapId,
+        handler: HandlerType.OnWebSocketEvent,
+        request: { method: '', params: { event } },
+      })
+      .catch((error) => {
+        logError(
+          `An error occurred while handling a WebSocket message for Snap "${snapId}":`,
+          error,
+        );
+      });
+  }
+
+  async open(snapId: SnapId, url: string, protocols?: string[]) {
+    assert(
+      !this.#exists(snapId, url),
+      `An open WebSocket connection to ${url} already exists.`,
+    );
+
+    const parsedUrl = new URL(url);
+    const { origin } = parsedUrl;
+
+    const id = nanoid();
+
+    // eslint-disable-next-line no-restricted-globals
+    const socket = new WebSocket(url, protocols);
+
+    const { promise, resolve, reject } = createDeferredPromise();
+
+    socket.addEventListener('open', () => {
+      resolve();
+      this.#handleEvent(snapId, {
+        type: 'open',
+        id,
+        origin,
+      });
+    });
+
+    socket.addEventListener('close', (event) => {
+      this.#handleEvent(snapId, {
+        type: 'close',
+        id,
+        origin,
+        code: event.code,
+        reason: event.reason,
+        wasClean: event.wasClean,
+      });
+    });
+
+    socket.addEventListener('error', () => {
+      reject(
+        rpcErrors.resourceUnavailable(
+          'An error occurred while opening the WebSocket.',
+        ),
+      );
+    });
+
+    // eslint-disable-next-line @typescript-eslint/no-misused-promises
+    socket.addEventListener('message', async (event) => {
+      const isText = typeof event.data === 'string';
+      const data = isText
+        ? { type: 'text' as const, message: event.data }
+        : {
+            type: 'binary' as const,
+            // We assume binary data to be sent via Blob for now.
```

Comment:
Can binary data be sent in other ways?

Reply:
Seems like the default is blobs: https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/binaryType

But we should verify that it works on mobile

Reply:
Might be, will check what works on mobile